spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java

```diff
@@ -159,16 +188,164 @@ public ListenableFuture<Job> writeDataToTable(
+      TableInfo tableInfo = TableInfo.newBuilder(tableId, tableDefinition).build();
+      bigQuery.create(tableInfo);
+    }
+  }
+
+  /**
+   * This method uses BigQuery Storage Write API to write new line delimited JSON file to the
+   * specified table. The Table should already be created as BigQuery Storage Write API doesn't
+   * create it automatically.
+   *
+   * @param tableName name of the table to write to
+   * @param jsonInputStream input stream of the json file to be written
+   * @return {@link ListenableFuture} containing the WriteApiResponse indicating completion of
+   *     operation
+   */
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream) {
+
+    SettableListenableFuture<WriteApiResponse> writeApiFutureResponse =
+        new SettableListenableFuture<>();
+    CompletableFuture<Void> completableFuture =
+        CompletableFuture.runAsync(
+            () -> {
+              try {
+                WriteApiResponse apiResponse = getWriteApiResponse(tableName, jsonInputStream);
+                writeApiFutureResponse.set(apiResponse);
+              } catch (DescriptorValidationException | IOException e) {
+                writeApiFutureResponse.setException(e);
+                logger.warn(String.format("Error: %s\n", e.getMessage()), e);
+              } catch (InterruptedException e) {
+                writeApiFutureResponse.setException(e);
+                // Restore interrupted state in case of an InterruptedException
+                Thread.currentThread().interrupt();
+              }
+            });
+
+    // register success and failure callback
+    writeApiFutureResponse.addCallback(
+        (res) -> {
+          completableFuture.cancel(true); // interrupt the completableFuture
+        },
+        (res) -> {
+          completableFuture.cancel(true); // interrupt the completableFuture
+        });
+
+    return writeApiFutureResponse;
+  }
+
+  private WriteApiResponse getWriteApiResponse(String tableName, InputStream jsonInputStream)
+      throws DescriptorValidationException, IOException, InterruptedException {
+    WriteApiResponse apiResponse = new WriteApiResponse();
+    BigQueryWriteClient client = BigQueryWriteClient.create();
+    TableName parentTable =
+        TableName.of(bigQuery.getOptions().getProjectId(), datasetName, tableName);
+
+    // Initialize a write stream for the specified table.
+    BigQueryJsonDataWriter writer = new BigQueryJsonDataWriter(parentTable, client);
+
+    try {
+      // Write data in batches. Ref: https://cloud.google.com/bigquery/quotas#write-api-limits
+      long offset = 0;
+      int currentBatchSize = 0;
+
+      BufferedReader jsonReader = new BufferedReader(new InputStreamReader(jsonInputStream));
+      String jsonLine = null;
+      JSONArray jsonBatch = new JSONArray();
+      while ((jsonLine = jsonReader.readLine()) != null) { // read the input stream line by line
```

Comment:
Will every line of input be required to be a valid JSON object? What about fixed-length formatted input, or minimized input in which all objects are on a single line?

Reply:
Do we want to restrict to newline delimited JSON format and then we can explore fixed-length formatted input, or minimised input as per the user's feedback. BQ Write API supports newLine Separated JSON out of the box (https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-json) and we may have to do additional processing at the client library layer for any additional formats we add, which we can plan to add as per the user's demand. WDYT?